### PR TITLE
Mangabz: rewrite, make multisrc, add mirror and new source

### LINKED ISSUE #12891: vomic
### Source name

vomic

### Source link

http://www.iewoai.com/#/home

### Language

Chinese

### Other details

This website has a lot of comic resources, it is recommended to make plug-ins, thank you 

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a title with source name.
- [X] I have checked that the extension does not already exist on the [website extensions list](https://tachiyomi.org/extensions/) or the app.
- [X] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/tachiyomiorg/tachiyomi-extensions/) and verified it does not appear in the code base.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by stevenyomi**
Rewrote Mangabz:
- Better parsing and preferences
- Added filters
- Reduce requests and cache results
- Updated launcher icons
- Added mirror to Mangabz
  - Xmanhua: https://xmanhua.com/
- Added source: vomic (2 mirrors), closes #12891
  - http://www.vomicmh.com/
  - http://www.iewoai.com/

Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

File: multisrc/overrides/mangabz/vomic/src/Vomic.kt [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                else -> url
+            }
+            Page(index, imageUrl = imageUrl)
+        }
+    }
+
+    override fun imageRequest(page: Page): Request {
+        val url = page.imageUrl!!
+        val host = url.toHttpUrl().host
+        val headers = headersBuilder().set("Referer", "https://$host/").build()
```
Comment (vetleledaal) [inline multisrc/overrides/mangabz/vomic/src/Vomic.kt:128]: is this correct? vomic appears to use `http://`

Comment (stevenyomi) [reply multisrc/overrides/mangabz/vomic/src/Vomic.kt:128]: This is an aggregator site and the image URLs are from other sites. The referer is set to access images on those sites' servers.

File: multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangabz/MangabzTheme.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    override fun latestUpdatesRequest(page: Int) = GET("$baseUrl/manga-list-0-0-2-p$page/", headers)
+
+    override fun latestUpdatesParse(response: Response) = searchMangaParse(response)
+
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList) =
+        if (query.isEmpty())
+            popularMangaRequest(page)
+        else
+            GET("$baseUrl/search?title=$query&page=$page")
```
[the remark this answers is not in the document]
Comment (stevenyomi) [reply multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangabz/MangabzTheme.kt:39]: This is safe for most queries in Chinese.

In fact, according to my test, okhttp will parse and escape any query parameter correctly even if I pass a query like `8* = ^562 我 | <`, so I guess it's completely safe to do this.

[the remark this answers is not in the document]
Comment (stevenyomi) [reply multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangabz/MangabzTheme.kt]: Nevermind, I forgot ampersands and percent symbols. Changed.

File: multisrc/overrides/mangabz/mangabz/src/CookieInterceptor.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+class CookieInterceptor(
+    private val domain: String,
+    private val key: String,
+    private val value: String
+) : Interceptor {
+
+    init {
+        val url = "https://$domain/"
+        val cookie = "$key=$value; Domain=$domain; Path=/"
+        CookieManager.getInstance().setCookie(url, cookie)
```
Comment (arkon) [inline multisrc/overrides/mangabz/mangabz/src/CookieInterceptor.kt:16]: The inspector that runs on push doesn't like that it's running a stubbed function here:

```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at suwayomi.tachidesk.manga.impl.util.PackageTools.loadExtensionSources(PackageTools.kt:123)
	at suwayomi.tachidesk.manga.impl.extension.Extension.installAPK(Extension.kt:53)
	at suwayomi.tachidesk.MainKt.main(Main.kt:46)
	at suwayomi.tachidesk.MainKt$main$2.invoke(Main.kt)
	at suwayomi.tachidesk.MainKt$main$2.invoke(Main.kt)
	at kotlin.coroutines.intrinsics.IntrinsicsKt__IntrinsicsJvmKt$createCoroutineUnintercepted$$inlined$createCoroutineFromSuspendFunction$IntrinsicsKt__IntrinsicsJvmKt$1.invokeSuspend(IntrinsicsJvm.kt:205)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:115)
	at kotlin.coroutines.jvm.internal.RunSuspendKt.runSuspend(RunSuspend.kt:19)
	at suwayomi.tachidesk.MainKt.main(Main.kt)
Caused by: kotlin.NotImplementedError: An operation is not implemented.
	at xyz.nulldev.androidcompat.androidimpl.StubbedCookieManager.setCookie(StubbedCookieManager.kt:25)
	at eu.kanade.tachiyomi.extension.zh.mangabz.CookieInterceptor.<init>(Unknown Source)
	at eu.kanade.tachiyomi.extension.zh.mangabz.Mangabz.<init>(Unknown Source)
	... 14 more
```

Comment (stevenyomi) [reply multisrc/overrides/mangabz/mangabz/src/CookieInterceptor.kt:16]: Let me wrap it with try-catch.

**Pull request merged**